### PR TITLE
Zgc/dipu optimize performance

### PR DESCRIPTION
**Pull request by zhaoguochun1995**
1. Reduced memory usage during model training by optimizing operators such as conv, linear, and cross_entropy
2. The autogen tool is enhanced to support operator fallback such as conv_backward and linear_backward and to perform autocompare
3. Made some optimizations to improve performance
4. Enhanced autocompare, operators that did not support fallback before also support autocompare (custom_fallback needs to be provided)
5. Fixed a bug that caused a segment error when recompiling dipu during model training
6. Fix timeline bug on nv
7. Added operator information grabber
[dipu_ops.xlsx](https://github.com/DeepLink-org/DIPU/files/12471241/dipu_ops.xlsx)
[dipu_ops_with_fallback_op_args.xlsx](https://github.com/DeepLink-org/DIPU/files/12494148/dipu_ops_with_fallback_op_args.xlsx)

### CODE REVIEW

File: torch_dipu/csrc_dipu/runtime/core/DIPUEventPool.cpp [hunk 1]
```diff
@@ -40,11 +40,11 @@ class EventPool final {
     }
 
     void get(T& event) {
-        std::lock_guard<mutex_t> _(event_mutex_);
         if (event_pool_.empty()) {
```
Comment (caikun-pjlab) [inline torch_dipu/csrc_dipu/runtime/core/DIPUEventPool.cpp:43]: event_pool_.empty()还是需要锁的保护

File: torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp [hunk 2]
```diff
@@ -66,20 +35,13 @@ class BFCachingAllocatorImpl{
...
-                    emptyCacheWithoutLock();
-                } else {
-                    throw std::runtime_error("no device memory available");
-                }
-            }
+        try {
+            ptr = allocate_fn(nbytes);
+            cachedBytes += nbytes;
+        } catch(...) {
+
```
Comment (caikun-pjlab) [inline torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:42]: allocate失败加点日志吧

File: torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp [hunk 3]
```diff
@@ -166,6 +128,7 @@ class BFCachingAllocatorImpl{
     std::vector<StreamSetHandle> streamSets_;
 
     using mutex_t = SpinMutex;
+    //using mutex_t = std::recursive_mutex;
```
Comment (caikun-pjlab) [inline torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:131]: 直接删除

File: torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp [hunk 4]
```diff
@@ -503,14 +466,17 @@ class BFCachingAllocator: public CacheAllocator {
...
-        block = impl->allocateRaw(size);
-    }catch(...) {
+    std::tuple<void*, int, size_t> block = impl->allocateRaw(size);
+    void* ptr = std::get<0>(block);
+    if (ptr == nullptr && size > 0) {
         empty_cache();
         block = impl->allocateRaw(size);
+        ptr = std::get<0>(block);
+        if (ptr == nullptr) {
+            throw std::runtime_error("no device memory available");
```
Comment (caikun-pjlab) [inline torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:476]: 推荐使用TORCH_CHECK，能够打印出堆栈，对定位问题更友好

File: torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp (hunk 2 above)
Comment (zhaoguochun1995) [reply torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:42]: 在后面，不在这里

File: torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp (hunk 4 above)
Comment (zhaoguochun1995) [reply torch_dipu/csrc_dipu/runtime/core/allocator/DIPUBFCachingAllocator.cpp:476]: 会打印出较多cached-allocator的调用栈，感觉没太必要

**Review by fandaoyi — APPROVED** [review]

**Review by caikun-pjlab — APPROVED** [review]

**Pull request merged**